### PR TITLE
Handle invalid JSON-LD gracefully

### LINKED ISSUE #31314: LinkCrawlWorker JSON::LD::JsonLdError::CollidingKeywords
### Steps to reproduce the problem

Observed in Sidekiq morgue after merging https://github.com/mastodon/mastodon/pull/31288

image

Errorr Class	JSON::LD::JsonLdError::CollidingKeywords
Error Message	colliding keywords: @id already exists in result

### Expected behaviour

No error

### Actual behaviour

Error

### Detailed description

Status that generated error https://mastodon.social/users/hessenschau/statuses/112915069178628443

### Mastodon instance

vmst.io

### Mastodon version

main

### Technical details

Mastodon	4.3.0-alpha.5+pr-31268
Ruby	3.3.4p94
PostgreSQL	16.3
Redis	7.2.4
OpenSearch	2.14.0
libvips	8.15.2
FFmpeg	7.0.2

### PR DESCRIPTION
**Pull request by c960657**
Fixes #31314 – ~though it should be discussed whether this should be fixed at all~.

### CODE REVIEW

File: app/lib/link_details_extractor.rb [hunk 1]
```diff
@@ -101,9 +101,17 @@ def root_array(root)
...
     def json
-      @json ||= root_array(Oj.load(@data))
-                .map { |node| JSON::LD::API.compact(node, 'https://schema.org') }
-                .find { |node| SUPPORTED_TYPES.include?(node['type']) } || {}
+      @json ||= begin
+        nodes = root_array(Oj.load(@data)).map do |node|
+          begin
+            JSON::LD::API.compact(node, 'https://schema.org')
+          rescue JSON::LD::JsonLdError
+            node['type'] = node['@type']
```
Comment (ClearlyClaire) [inline app/lib/link_details_extractor.rb:109]: Why this change?

Comment (c960657) [reply app/lib/link_details_extractor.rb:109]: Because the compacting converts `@type` to `type`. I don't understand how compacting works – I just assume this is intended. 

Prior to #31288 we were checking `@type`: https://github.com/mastodon/mastodon/commit/e9e63ed203996a4e06e31a7220e9d8a2df71f67c

Comment (ClearlyClaire) [reply app/lib/link_details_extractor.rb:109]: Note that we have decided to revert JSON-LD compacting for now. I'll explain in more details when I have more time, but basically it was creating way more issues than it was solving (and did not fix the original issue, which as not the presence of *a* named graph, but multiple ones)

**Comment by c960657** [comment]
JSON-LD parsing was reverted in #31355, closing.

File: app/lib/link_details_extractor.rb (hunk 1 above)
Comment (ClearlyClaire) [reply app/lib/link_details_extractor.rb:109]: So, to expand a bit on this:
- the issue we tried to solve using JSON-LD compacting was a document having multiple named graphs, but compacting by itself does not help, we need to select the graph we are interested in; compacting would have worked if there was only one named graph but that was not the case
- compacting JSON-LD requires knowing the full context, which may require fetching remote documents; in particular, the `json-ld` gem will by default automatically fetch remote documents with its own networking code (that, in particular, does not have our timeouts or address range checks)
- the most correct approach would be to expand the JSON-LD document and check the exact properties we are interested in, but just as with compacting, this can require fetching remote documents, and is furthermore very brittle because it will not handle malformed documents well even though those can generally be assumed to have the expected shape when considering them as simple JSON
